server/svix-server/src/core/types.rs

```diff
@@ -490,6 +490,42 @@ pub struct ExpiringSigningKey {
     pub expiration: DateTime<Utc>,
 }
 
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct EndpointHeaders(pub HashMap<String, String>);
+json_wrapper!(EndpointHeaders);
+
+impl<'de> Deserialize<'de> for EndpointHeaders {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        HashMap::deserialize(deserializer)
+            .map(|x: HashMap<String, String>| {
+                x.into_iter().map(|(k, v)| (k.to_lowercase(), v)).collect()
+            })
+            .map(EndpointHeaders)
+    }
+}
+
+impl Validate for EndpointHeaders {
+    fn validate(&self) -> std::result::Result<(), ValidationErrors> {
+        let mut errors = ValidationErrors::new();
+        self.0.iter().for_each(|(k, v)| {
+            if let Err(_e) = http::header::HeaderName::try_from(k) {
```

Comment:
Not saying one way or the other, but why use `HeaderName` only in the validation here, and not also in the type of the `HashMap` used in `EndpointHeaders`?

As in, why not change the above to:
```
 pub struct EndpointHeaders(pub HashMap<http::header::HeaderName, http::header::HeaderValue>);
```

Not saying one way is better than the other, I was just wondering.

Reply:
OK, we chatted off-band, let's leave it like this for now.

Reply:
Sorry meant to post a comment here earlier -- I agree that either using the `http::header` types here are a better long-term solution. However, I've run into some scenarios using them where I can cause thread panics. I think it makes sense to leave current implementation in place until I can find a way to cleanly avoid unwanted panics.